services/pagerduty2/service.go

```diff
@@ -0,0 +1,388 @@
+//
+// The req headers are now required with the API v2:
+// https://v2.developer.pagerduty.com/docs/migrating-to-api-v2
+func (s *Service) Alert(serviceKey, incidentKey, desc string, level alert.Level, details string, timestamp time.Time, data alert.EventData) error {
+	url, post, err := s.preparePost(serviceKey, incidentKey, desc, level, details, timestamp, data)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", url, post)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/vnd.pagerduty+json;version=2")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusAccepted {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Printf("error parsing error body\n")
+			return err
+		}
+		type response struct {
+			Message string `json:"message"`
+		}
+		r := &response{Message: fmt.Sprintf("failed to understand PagerDuty2 response. code: %d content: %s", resp.StatusCode, string(body))}
+		b := bytes.NewReader(body)
+		dec := json.NewDecoder(b)
+		dec.Decode(r)
+		return errors.New(r.Message)
+	}
+	return nil
+}
+
+func (s *Service) sendResolve(c Config, serviceKey, incidentKey string) (string, io.Reader, error) {
+	// create a new AlertPayload for us to fire off
+	type Resolve struct {
+		RoutingKey  string `json:"routing_key"`
+		DedupKey    string `json:"dedup_key"`
+		EventAction string `json:"event_action"`
+	}
+
+	ap := Resolve{}
+
+	if serviceKey == "" {
+		ap.RoutingKey = c.ServiceKey
+	} else {
+		ap.RoutingKey = serviceKey
+	}
+
+	ap.DedupKey = incidentKey
+	ap.EventAction = "resolve"
+
+	var post bytes.Buffer
+	enc := json.NewEncoder(&post)
+	err := enc.Encode(ap)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return c.URL, &post, nil
+}
+
+// preparePost is a helper method that sets up the payload for transmission to PagerDuty
+func (s *Service) preparePost(serviceKey, incidentKey, desc string, level alert.Level, details string, timestamp time.Time, data alert.EventData) (string, io.Reader, error) {
+	c := s.config()
+	if !c.Enabled {
+		return "", nil, errors.New("service is not enabled")
+	}
+
+	var severity string
+	eventType := "trigger"
+
+	switch level {
+	case alert.Warning:
+		severity = "warning"
+	case alert.Critical:
+		severity = "critical"
+	case alert.Info:
+		severity = "info"
+	default:
+		// default is a 'resolve' function
+		return s.sendResolve(c, serviceKey, incidentKey)
+	}
+
+	// create a new AlertPayload for us to fire off
+	ap := &AlertPayload{
+		Payload: &PDCEF{},
+	}
+
+	if serviceKey == "" {
+		ap.RoutingKey = c.ServiceKey
+	} else {
+		ap.RoutingKey = serviceKey
+	}
+
+	ap.Client = "kapacitor"
+	ap.ClientURL = s.HTTPDService.URL()
+	ap.DedupKey = incidentKey
+	ap.EventAction = eventType
+
+	ap.Payload.CustomDetails = make(map[string]interface{})
```

Comment:
I think it would be simpler and still just as useful if the CustomDetails where just the data.Result object. The Result object serializes to JSON in a standard format and contains all relevant data to the alert. Will it still work for your use case to just ignore the `details` string and pass in the Result directly? This way we can avoid any edge cases with trying to unmarshal the details strings since its opaque to the code at this point. Thoughts?

To be clear I am thinking something like this

```
ap.Payload.CustomDetails = make(map[string]interface{})
ap.Payload.CustomDetails["result"] = data.Result
```

Since the details strings can only be constructed using the data already in data.Result, then this is a superset of the data being provided. The only question is if the data is still useful if we do it this way.

Reply:
Ah, that makes sense. WIll update and get back to you ASAP!